pm: policy: extend and improve API

gmarull opened the pull request:
This PR contains a series of improvements and additions to the policy API. They can be summarized by:

- Rename "residency" and "app" policy naming to "default" and "custom", respectively. Since Zephyr comes with a policy that takes into account residency times as well as other constraints, the residency naming alone may be misleading. 
- Moved constraints API to policy, renaming it to the policy namespace and with a more descriptive naming, e.g. `pm_policy_state_lock_put`. The constraint API was only used by the policy manager, so it was not useful used alone. New tests have been added that check the behavior of the policy manager when certain states are allowed or not.
- Added a new set of policy APIs to impose maximum latency. This can be useful, for example, if drivers or application need fast response times and do not want to be impacted by certain PM states.
- Multiple cleanups: removed redundant includes, logging, assertions, etc.

gmarull committed:
pm: policy: use DEFAULT/CUSTOM naming

The residency policy, is in reality, influences by other parameters for
example constraints. It has been renamed to "DEFAULT" policy to make it
more general. The "APP" policy has been renamed to "CUSTOM" to better
represent its purpose.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: move to pm folder

Move policy code one level up for simplicity, since there is a single
source file. Source file has been renamed to "policy.c" to make things
clear.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: remove unused logging

The policy file does not perform any logging, so just drop logging
include and declaration.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: constraint: remove redundant assertions

1. state < PM_STATE_COUNT: the function argument is an enumeration that
   already indicates the valid values. It is a programming error to pass
   random integer values, the API should not be defensive "just in case
   the application has messed things up".
2. v < UINT_MAX: first, this assumes type used by `atomic_t` (opaque).
   Second, if the integer overflows it will just wrap around, so the
   check does nothing in practice.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: constraint: remove unused logging

The constraint code does not log, so remove include/module declaration.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: constraint: cleanup include list

Most include files were not used, and `toolchain.h` was missing (needed
by `__weak`).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: move constraints to policy API

The pm_constraint_* APIs were effectively used by the policy manager
only. This patch renames the API to the policy namespace and makes its
naming more explicit:

- pm_constraint_set -> pm_policy_state_lock_get()
- pm_constraint_release -> pm_policy_state_lock_put()
- pm_constraint_get -> pm_policy_state_lock_is_active()

The reason for these changes is that constraints can be of many types:
allow/disallow states, impose latency requirements, etc. The new naming
also makes explicit that the API calls will influence the PM policy
behavior.

All drivers and documentation have been updated accordingly.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: i2c: cc13xx_cc26xx: remove CONFIG_PM guards

API provides no-op fallback when CONFIG_PM=n

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: spi: cc13xx_cc26xx: remove CONFIG_PM guards

API provides no-op fallback when CONFIG_PM=n

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: pm: policy_api: check status lock behavior

Test that policy manager behavior is correct when allowing/forbidding
certain states.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: assert on unbalanced state lock get/put

The state lock get/put calls should always be balanced: first call get
and then put. Add an assertion in case lock counter is about to go
negative, indicating a programming error.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: new API to set maximum latency requirements

Add a new API that allows to configure maximum latency requirements.
When the policy manager computes the next state, it will check if the
state brings too much latency based on requirements. This can be useful,
for example, if a certain driver or the application want a system to
respond fast, since any low power state that brings too much latency
will not be used.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: pm: policy_api: test latency API

Add a new tests that checks the behavior of the policy when latency
requirements are imposed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: allow to get notifief when maximum latency changes

Add a new API that allows to register a callback that is called when the
maximum latency value changes. This can be used by SoC code to perform
actions based on certain latency values.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: pm: policy_api: test latency change callback

Test that we get notified correctly when latency requirements change.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

pdunaj commented on drivers/serial/uart_ns16550.c:
```diff
@@ -681,7 +681,7 @@ static void uart_ns16550_irq_tx_enable(const struct device *dev)
 		 */
 		for (uint8_t i = 0U; i < num_cpu_states; i++) {
-			pm_constraint_set(cpu_states[i].state);
+			pm_policy_state_disallow(cpu_states[i].state);
```
how they can restrict all states? PM_STATE_ACTIVE is at zero, so what state will be allowed to be in if all are not allowed?

gmarull replied:
`PM_STATE_ACTIVE` is effectively ignored (policy will return `NULL`, indicating CPU must remain active).

pdunaj replied:
Would it be the same if disallow will come when other state is set?
Adding short info in function api doc would be useful.
e.g.
```
Request to disallow all states will result in CPU staying in active state, until restrictin is lifted.
```

gmarull replied:
clarified

pdunaj commented on include/pm/policy.h:
```diff
@@ -7,13 +7,46 @@
+ *
+ * @param latency New maximum latency in microseconds.
+ */
+typedef void (*pm_policy_latency_changed_cb_t)(int32_t latency);
```
why latency is signed int? are negative values supported - if so what these mean?

gmarull replied:
it is signed because we also need `SYS_FOREVER_US` (which is -1)

pdunaj replied:
let's add small info
```
Positive value represents latency in microseconds. Negative value of -1 lifts the latency constraint. Other values are forbidden.
```

Maybe `k_timeout_t` can be used here?

gmarull replied:
clarified range of values. I'll check more in detail some existing opaque types and their usage (e.g. `k_ticks_t`, `k_timeout_t`, etc.).

pizi-nordic replied:
I would not bound this value to ticks, as there is a higre difference between 50ns and 50us of latency in some cases but all the values are well within one tick.

gmarull replied:
> I would not bound this value to ticks, as there is a higre difference between 50ns and 50us of latency in some cases but all the values are well within one tick.

Not sure if I'm following, latency is specified in `us`, not ticks.

ceolin replied:
Agree

pdunaj commented on include/pm/policy.h:
```diff
@@ -33,6 +66,102 @@ const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks);
+ * pm_policy_state_disallow(). Subsequent calls will just increase a reference
+ * count, thus meaning this API can be safely used concurrently. A state will
+ * be allowed again after pm_policy_state_allow() is called as many times as
+ * pm_policy_state_disallow().
```
considering one of the above examples - what will happen when user disallows all?

gmarull replied:
clarified docs

pdunaj commented on include/pm/policy.h:
```diff
@@ -33,6 +66,102 @@ const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks);
+ * @param value Maximum allowed latency in microseconds.
+ */
+void pm_policy_latency_request_add(struct pm_policy_latency_request *req,
+				   uint32_t value);
```
why explicit bitwidth (not simply unsigned int)?

pdunaj commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+/** Maximum CPU latency in ticks */
+static int32_t max_latency_ticks = K_TICKS_FOREVER;
+/** Callback to notify when maximum latency changes. */
+pm_policy_latency_changed_cb_t latency_changed_cb;
```
static

pdunaj commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+	struct pm_policy_latency_request *req;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&latency_reqs, req, node) {
+		if ((new_max_latency_ticks == K_TICKS_FOREVER) ||
```
maybe we could keep the list sorted for better performance?

gmarull replied:
I'll take a look if it's worth, since this operation is `O(N)`, doing a linear search to insert/update is also `O(N)`. Removal would benefit, though.

pizi-nordic replied:
If we keep list sorted (in order of increased latency), then the stricter requirements will be faster to insert/remove. Note, that our interest lays in getting the most strict value.

gmarull replied:
If we keep the list sorted, we still need to do a search when inserting, which I guess the best we can have is `O(log N)` by partitioning. I'd first wait to see performance numbers before going into micro-optimizations.

ceolin replied:
Agree ... How many latencies we expect to have ? Probably not much that justify
the complexity. Also, this is a linked list, which means that even to do a
binary search there is an overhead to find the middle element and you have to
have access to the last element. So this probably only justify for big lists.

pdunaj commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+
+	num_cpu_states = pm_state_cpu_get_all(cpu, &cpu_states);
+
+	for (int16_t i = (int16_t)num_cpu_states - 1; i >= 0; i--) {
```
why int16_t?

gmarull replied:
`num_cpu_states` is `uint8_t`, we need to promote here (`0` case)

pdunaj replied:
maybe int? I personally avoid using specific widths unless needed so compiler can chose the best one.

gmarull replied:
Since we know the necessary bit width here, promoting to `int16_t` is the right thing to do IMO.

pdunaj commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+{
+	atomic_t cnt = atomic_dec(&state_disallow_cnt[state]);
+
+	__ASSERT(cnt >= 1, "Unbalanced state disallow/allow");
```
but why is it not allowed? what if two modules disallow at the same time and later one of them lifts the restriction?

if the assert is left only one module would be able to keep the restriction at any given time which seems like a huge limitation.

gmarull replied:
Calling `allow` without a previous `disallow` is wrong, so an unbalanced sequence should never happen in well-behaved code.

pdunaj replied:
Ah... yes. I read it the wrong way around. For some reason though this is restricting multiple disallow.
It is ok obviously :) Sorry

gmarull commented on drivers/serial/uart_npcx.c:
```diff
@@ -64,19 +64,19 @@ struct uart_npcx_data {
-	if (atomic_test_and_set_bit(data->pm_constraint_flag, flag) == 0) {
-		pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
+	if (atomic_test_and_set_bit(data->pm_policy_state_flag, flag) == 0) {
+		pm_policy_state_disallow(PM_STATE_SUSPEND_TO_IDLE);
```
Agreed that implementation has to be fixed/improved, but I think we should address these issues as part of new PRs.

pdunaj reviewed (commented):
seems ok

pdunaj commented:
Hi @gmarull , can you force push review updates and PR rebases separately? It makes it easier to see the changes (via the force-pushed comparison) :)

gmarull commented:
> Hi @gmarull , can you force push review updates and PR rebases separately? It makes it easier to see the changes (via the force-pushed comparison) :)

sure, will do

ceolin commented on soc/arm/ti_simplelink/cc13x2_cc26x2/power.c:
```diff
@@ -184,57 +208,5 @@ void PowerCC26XX_schedulerRestore(void)
-	return ret;
-}
-#endif /* CONFIG_PM */
-
```
As I've mentioned I think that this change is risky. Removing the capability of a target implement their own constraint and just let them be notified when a constraint is set or release is not enough if the target set constraints not using Zephyr's API. Taking as example the TI HAL, In https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/ti_simplelink/cc13x2_cc26x2/power.c#L156-L168 we call `Power_Init()` that sets a constraint using the HAL API https://github.com/zephyrproject-rtos/hal_ti/blob/master/simplelink/source/ti/drivers/power/PowerCC26X2.c#L366-L414

pizi-nordic commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+/** State disallow reference counting */
+static atomic_t state_disallow_cnt[PM_STATE_COUNT];
+/** Callback to notify when state allowed status changes */
+static pm_policy_state_allowed_changed_cb_t allowed_changed_cb;
```
What if several modules have to be notified about policy/latency change?
It looks that we can have only one registered callback, but exported API is supposed to be used by multiple clients.

gmarull replied:
This is intended for SoC-specific usage (that's the requirement I got). If we want to have a list of clients, we need to clarify the use case.

pizi-nordic commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+static struct k_spinlock latency_lock;
+/** List of maximum latency requests. */
+static sys_slist_t latency_reqs;
+/** Maximum CPU latency in ticks */
```
Tick is too large unit. I would prefer having latency expressed in us (realistic) or ns (future-proof).

gmarull replied:
This is an internal unit, not the units used by public API.

pizi-nordic replied:
Still, this unit determines timing resolution of the whole solution.

gmarull replied:
Unfortunately we get ticks, so it's the best we can do right now.

pizi-nordic commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+			latency_us = (int32_t)k_ticks_to_us_ceil32(new_max_latency_ticks);
+		}
+
+		latency_changed_cb(latency_us);
```
User-provided callback called with spinlock held might be a recipe for lockup.

gmarull replied:
We can't offer safety guarantee otherwise: another user may call for a change before this completed.

pizi-nordic commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+		total_latency = min_residency + exit_latency;
+
+		/* skip state if it brings too much latency */
+		if ((max_latency_ticks != K_TICKS_FOREVER) &&
```
That is not always true, as we can handle interrupts before we enter sleep mode. So here exit_latency is important.

gmarull replied:
Please clarify a bit more here

pizi-nordic replied:
To make decisions you are just summing min_residency and exit_latency. Here, only exit_latency should be used.

gmarull replied:
Right, fixed thanks for catching.

pizi-nordic commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+			continue;
+		}
+
+		if ((ticks == K_TICKS_FOREVER) || (ticks >= total_latency)) {
```
Where we are checking expected residency time?

gmarull replied:
Please clarify a bit more here

pizi-nordic replied:
Expected residency time is the time which we are supposed to spend in sleep mode. I do not see any check taking that part under consideration.

gmarull replied:
this is given in `ticks` (policy is called with this value obtained from the Kernel)

pizi-nordic commented on subsys/pm/policy.c:
```diff
@@ -0,0 +1,165 @@
+	return (atomic_get(&state_disallow_cnt[state]) == 0);
+}
+
+void pm_policy_state_allowed_changed(pm_policy_state_allowed_changed_cb_t cb)
```
Same applies for latency changed callback.

gmarull committed:
manifest: update hal_ti

Update hal_ti to pull updates on the PM constraints API.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin commented on include/pm/pm.h:
```diff
@@ -115,53 +115,15 @@ int pm_notifier_unregister(struct pm_notifier *notifier);
+ * @param event_us Event time in microseconds.
  */
-bool pm_constraint_get(enum pm_state state);
+void pm_event_hint(uint32_t event_us);
```
This is more than a hint IMHO. It is adding a timeout that will
trigger an interruption much likely waking up the SoC (or preventing
it from sleep). There is not much what the policy can do with it right
? Since the policy can distinguish this "hint" timeout from any other
scheduled event.

gmarull replied:
This should influence the system because the next scheduled event may be the one provided by the hint, so, the policy will get that (in ticks) effectively influencing the next possible states. Do you have any other name suggestion?

ceolin replied:
Yes, it will influence but more thant I hint IMHO. This will effectivelly trigger an event in the system (the way it is implemented). I thought it would just be used by the policy to influence which power state to select. Just affraid that it ends up causing an overhead triggerning many interruptions.

gmarull replied:
@carlocaione can you review this?

carlocaione replied:
> it ends up causing an overhead triggerning many interruptions.

Why is that? you are literally informing the system that a new event will be likely to happen once in the future and to take that into account. I don't fully understand the concern for overhead here.

ceolin replied:
My point is if should just be used by the policy to decide if it wants to wake up  earlier (or not even supspend) or not. Setting a timeout the policy does not know that this is a "hint". Basically you are enforcing an interruption to happen.
When I see a "hint" I interpret as an information that the policy can use to decide what is the best state to use but that is not case here. since the policy can't distinguish between this "hint" and a real event.

ceolin commented on include/pm/policy.h:
```diff
@@ -33,6 +59,121 @@ const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks);
+ */
+void pm_policy_state_lock_put(enum pm_state state);
+
+/**
```
Wouldn't be better ?
```
pm_policy_state_lock()
pm_policy_state_unlock()
pm_policy_state_is_locked()
```

gmarull replied:
I also thought the same, but I think `pm_policy_state_lock`/`pm_policy_state_unlock` doesn't emphasize that it's a reference counted API. Calling `pm_policy_state_unlock` doesn't necessarily mean the state is unlocked. The terminology used in this proposal is:

- pm_policy_state_lock_put: release (put) _state lock_
- pm_policy_state_lock_get: obtain (get) _state lock_

put/get terminology is re-used from PM runtime.

ceolin replied:
That is a good point ... lets move with this names then. The one that I really disliked is
`pm_policy_state_lock_is_active` :/ But it is better to keep the namespace consistent.

ceolin commented:
@gmarull the only point in these changes that I'm not super confdent is `pm_event_hint`, all other changes look great to me.

carlocaione commented:
> @gmarull the only point in these changes that I'm not super confdent is `pm_event_hint`, all other changes look great to me.

Are you not confident on the name so you want that named differently, or you are not confident on the actual usage of the function?

gmarull commented:
Should I remove the last commit to get this in so we can continue the discussion on `pm_event_hint`?

gmarull commented:
@carlocaione @ceolin I have dropped the `pm_event_hint` and rebased, we can continue the discussion in a new PR so we can move on. Can you also ack https://github.com/zephyrproject-rtos/hal_ti/pull/31 ?

gmarull commented:
@ceolin @carlocaione updated manifest SHA, should be ready

gmarull commented:
rebased (unrelated CI failures, hopefully fixed in main)

gmarull commented:
@ceolin ping

gmarull commented:
@carlocaione @ceolin @erwango please re-approve, rebased due to conflicts in a modified driver.

carlocaione reviewed (approved):
This PR is perpetually approved :D

gmarull commented:
> This PR is perpetually approved :D

This re-approval thing makes no sense at all. It's just frustrating.

erwango commented:
> > This PR is perpetually approved :D
> 
> This re-approval thing makes no sense at all. It's just frustrating.

For authors, it's just like building card castles.
And for reviewers, this begins to be counterproductive as now I found myself trying to limit the level of my remarks on PRs that are in a review for a long time and are starting to get approvals.

gmarull commented:
FYI https://github.com/zephyrproject-rtos/zephyr/issues/43701
